feat: add a producer id to BestSolutionChangedEvent

Christopher-Chianelli opened the pull request.

Christopher-Chianelli committed:
feat: add a producer id to BestSolutionChangedEvent

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/phase/custom/DefaultCustomPhase.java:
```diff
@@ -23,6 +23,7 @@
 public final class DefaultCustomPhase<Solution_>
         extends AbstractPossiblyInitializingPhase<Solution_>
         implements CustomPhase<Solution_> {
+    public static final String CUSTOM_STRING = "Custom";
```
I'd recommend adding a suffix "Phase" here, because "Custom" on its own doesn't say much.

triceo reviewed (commented):
Good idea on the event type! I think we can still improve on it.

Christopher-Chianelli committed:
chore: review comment

Christopher-Chianelli committed:
feat: partitioned search support

Christopher-Chianelli committed:
chore: sonar issues

triceo commented on core/src/main/java/ai/timefold/solver/core/api/solver/event/BestSolutionChangedEvent.java:
```diff
@@ -58,6 +70,15 @@ public long getTimeMillisSpent() {
+    /**
+     * @return A {@link EventProducerId} identifying what generated the event, either a
+     *         {@link SolveEventProducerId} if the cause is not associated with a Phase,
+     *         or {@link PhaseEventProducerId} if it is.
```
Please switch the order - start with the case that doesn't require negation; the sentence becomes easier to read.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/constructionheuristic/DefaultConstructionHeuristicPhase.java:
```diff
@@ -45,7 +48,12 @@ public TerminationStatus getTerminationStatus() {
 
     @Override
     public String getPhaseTypeString() {
```
Since `Phase` is not public API, we have some freedom here.
If you decide to keep `PhaseType` (see above), maybe we return that here, instead of the String? If we ever need to exhaustively switch over phase types, this will help.

triceo reviewed (commented):
I'll have to ask the platform guys about serialization of the events. That will give us some context here.

triceo commented on core/src/main/java/ai/timefold/solver/core/api/solver/event/EventProducerId.java:
```diff
@@ -0,0 +1,86 @@
+     * @return The index of the Phase that produced the event, or {@link OptionalInt#empty()} if the event producer
+     *         is not associated with a phase.
+     */
+    OptionalInt eventPhaseIndex();
```
Having asked the platform folks, we now know that they do their own serialization; therefore we needn't worry about this aspect. Still, I have a feeling that it'd work better as a single type. (Even though the platform does their own serialization, other users of our API may not; this API has been around since forever, there will be other users.)

Christopher-Chianelli committed:
chore: review comments

Christopher-Chianelli committed:
fix: flaky tests

A Problem Change does not trigger a new best solution event if it occurs
before Construction Heuristic finishes.

Christopher-Chianelli committed:
chore: add back phaseIndex

triceo reviewed (commented):
We're missing the identification on the first initialized solution consumer. After that, we're good.

Christopher-Chianelli committed:
chore: use events in the SolverJob consumers instead of passing the solution

Having a consumer that accepts the solution was fragile, since whenever
we wanted to add additional info, we would need to create an additional
overload.
Using events allows us to add new methods without affecting user code.

triceo reviewed (approved):
Only one comment. As agreed, I'll fix it myself.

triceo committed:
Review comments
